From 6e7b110a6b522cbd3fbba3e06d798972179cba24 Mon Sep 17 00:00:00 2001
From: Pradyot Bhattacharyya <pbhattacharyya@habana.ai>
Date: Wed, 05 Mar 2025 07:42:24 +0200
Subject: [PATCH] [FSW-14082] arcboot: fix ipas issues for bootrom

The following issues have been addressed:
boot_rom_is_valid_address_range() takes an unsigned
value into a signed parameter
There is no upper bound check on max number of
image pointers in a given manifest.

Change-Id: I29eaad40cbc459fce958b16cb44fc675b4bd32f5
---

diff --git a/arch/fs1/arc/common.c b/arch/fs1/arc/common.c
index afae35f..adf24b6 100644
--- a/arch/fs1/arc/common.c
+++ b/arch/fs1/arc/common.c
@@ -417,7 +417,7 @@
 }
 
 /* This function is only applicable for io die master */
-bool boot_rom_is_valid_address_range(uint64_t addr, int len)
+bool boot_rom_is_valid_address_range(uint64_t addr, uint32_t len)
 {
 	if (addr >= (uint64_t)SCRATCHPAD_SRAM_BASE &&
 	    addr + len <= (uint64_t)(SCRATCHPAD_SRAM_BASE + SCRATCHPAD_SRAM_SIZE))
diff --git a/common/boot_desc.c b/common/boot_desc.c
index d440120..701f542 100644
--- a/common/boot_desc.c
+++ b/common/boot_desc.c
@@ -145,6 +145,11 @@
 	p_offset = (uint32_t *)(p_spi_manifest_const->dest_addr_arc +
 				p_boot_ptr->offset2array);
 
+	if (p_boot_ptr->count > HL_BOOT_INFO_DESC_MAX_COUNT)
+	{
+		return BOOT_ROM_FAIL;
+	}
+
 	/* Validate all image descriptors */
 	for (i = 0; i < p_boot_ptr->count; i++) {
 		p_img_desc = (struct hl_fw_img_desc *)(p_spi_manifest_const->dest_addr_arc +
diff --git a/common/common.h b/common/common.h
index b38dc2a..8b34630 100644
--- a/common/common.h
+++ b/common/common.h
@@ -18,7 +18,7 @@
 enum boot_rom_status boot_rom_read_config_fuse(void);
 void boot_rom_set_efuse_read(void);
 void boot_rom_set_spi_efuse_read(void);
-bool boot_rom_is_valid_address_range(uint64_t addr, int len);
+bool boot_rom_is_valid_address_range(uint64_t addr, uint32_t len);
 void boot_rom_set_tpm_extend_done(void);
 enum boot_rom_tpm_exec_stage boot_rom_get_tpm_exec_stage(void);
 uint8_t boot_rom_get_tpm_next_pcr_index(void);
